### PR TITLE
errors: Is-method matching, absence assertions, and comparison helpers

### PR DESCRIPTION
**Pull request by amery**
## Summary

This branch widens error matching, fills gaps in the assertion
surface, and adds value-comparison helpers, all against the Go 1.25
floor. Highlights for consumers:

- `IsError` now recognises wrapped standard sentinels and any error
  carrying its own `Is(error) bool` method, not just identity matches.
- New absence and predicate/type assertions (`AssertNotErrorIs`,
  `AssertErrorIsFn`, `AssertErrorAs`) plus their `AssertMust*` twins.
- New `AreEqual` / `AreComparable` helpers that compare by `==`
  semantics without ever panicking, and report unknown rather than
  guessing.

## Error matching and sentinels

- **match errors by their own Is method in IsError** — each chain
  node now gets the same per-node test `errors.Is` applies: identity
  first, then the node's own `Is` method. Identity leads because some
  standard errors do not self-match (`syscall.Errno`), and the check
  is recover-guarded so a shared non-comparable dynamic type degrades
  to no-match instead of panicking. Exports `NewCheckErrorIsIn` and
  `NewCheckErrorIsIn2` for composing the same matching with
  `IsErrorFn` / `IsErrorFn2`, and retires the cognitive-complexity
  waivers on those functions.
- **alias ErrInvalid to fs.ErrInvalid** — the message text is
  identical (`"invalid argument"`), so aliasing changes no output
  while letting `errors.Is` match across the boundary with
  `fs.ErrInvalid` and `os.ErrInvalid` without help from `IsError`.

## Assertions

- **AssertNotErrorIs** (+ `AssertMustNotErrorIs`) — passes when
  `errors.Is` finds no match in the chain. A nil error passes against
  any non-nil target and fails against a nil target, mirroring
  `errors.Is(nil, nil)`.
- **AssertErrorIsFn and AssertErrorAs** (+ `AssertMust*` twins) —
  the first checks a chain against a predicate via `IsErrorFn`; the
  second extracts a typed error via `errors.As`, returning a pointer
  to the match. Constraining the type parameter to `error` turns a
  non-error type argument into a compile error instead of a runtime
  `errors.As` panic.

## Comparison helpers

- **AreEqual and AreComparable** — `AreComparable` answers whether
  `==` can panic on the given values; `AreEqual` reports whether each
  value equals the next as `==` would decide, without ever panicking.
  Same-type comparable values use `==`; typed nils, identity, and a
  reflection-probed, panic-guarded `Equal(T) bool` stand in where
  `==` is unavailable. Anything undecided is reported as unknown
  through the second result rather than resolved by deep comparison.
- **compare slices element by element, one level deep** — slices
  left undecided by nil, identity, or a decisive `Equal` method are
  walked element by element (lengths must match, interface elements
  unwrapped first). Nested slices are not walked — they settle only
  by nil, identity, or their own `Equal`, else stay unknown. A
  decisive `Equal` keeps precedence; a panicking or wrongly-signed
  one falls through to the walk. A nil slice and an empty one remain
  unequal, unlike `slices.Equal`.
- **guard IsSame against reflect.Values that cannot be unwrapped** —
  the primitive and interface arms of `isSamePointer` called
  `Interface()` unconditionally, panicking on values reached through
  unexported struct fields. Such values are now reported as not same
  (unknown through `AreEqual`'s fallback). Also rewords the IsSame
  doc so the composite-value-types exclusion states its rationale —
  sameness means a change would reach both operands, which distinct
  aggregates cannot guarantee — and points at `AreEqual` for
  by-value comparison.

## Preconditions and build

- **pin MustNoError\* edge-case behaviour** — regression tests and
  doc notes locking in typed-nil handling, the bare-sentinel
  no-cause panic shape, and inert nil entries in the allowed list.
  No behaviour change.
- **anchor cspell ignorePaths at any depth** — the config lives in
  `internal/build/`, so bare patterns resolved relative to that
  directory; prefixing them with `**/` applies them wherever the
  walk starts.

## Testing

`make all test` is green on the Go 1.25 floor. Every new function
carries table-driven or MockT-driven tests; `README.md`, `TESTING.md`
and `TESTING_core.md` are updated to list the additions.

## Summary by CodeRabbit

* **New Features**
  * Added `AreComparable` and `AreEqual` for panic-safe equality checks.
  * Expanded error-matching assertions with predicate-based, negated, type-extraction, and “must” variants, plus `NewCheckErrorIsIn`/`NewCheckErrorIsIn2`.
* **Bug Fixes**
  * Updated `ErrInvalid` to be an alias of `fs.ErrInvalid` for consistent `errors.Is` matching.
* **Documentation**
  * Refreshed README and testing docs with clarified comparison and error-chain semantics.
* **Tests**
  * Added coverage for typed-nil handling, wrapped errors, non-comparable values, and panic/abort behaviour.
* **Chores**
  * Updated spellcheck ignore patterns.

### CODE REVIEW

File: zero.go [hunk 1]
```diff
@@ -337,14 +344,274 @@ func isSamePointer(va, vb reflect.Value) bool {
...
+
+// isSameInterface extracts the values held by two interfaces and
+// compares them recursively. Interfaces that cannot be unwrapped —
+// unexported struct fields reached by reflection — are never same.
+func isSameInterface(va, vb reflect.Value) bool {
+	if !va.CanInterface() || !vb.CanInterface() {
+		return false
+	}
+	return IsSame(va.Elem().Interface(), vb.Elem().Interface())
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply zero.go:365]: `.IsValid()` is checked before calling this helper so `Elem()` can't return `nil`

**Comment by amery** [comment]
Branch updated (rebased on `main`, force-pushed) to address the
latest review feedback. Net change is small — one README line and a
pair of test rows — since one finding turned out to be already
guarded.

**Finding 1 — `isSameInterface` nil-interface panic: verified
unreachable, pinned by test rather than code.**
The mechanism is real in isolation: `Elem()` on a nil interface
returns the zero `reflect.Value`, and `.Interface()` on that panics.
But `isSameInterface` is only reached via `isSamePointer`, and both
callers (`isReflectValueSame`, `areEqualFallback`) run
`isSameTypedNil` first — which detects a nil interface through
`isReflectValueNil`'s `reflect.Interface` arm and decides it before
any unwrap. A guard inside `isSameInterface` would be dead code.
Instead I added two rows to `isSameInterfaceReflectValueTestCases`
using nil interface reflect.Values (`CanInterface() == true`,
`IsNil() == true`) — the exact feared state — proving `IsSame`
returns without panicking. That pins the reachability so a later
refactor of the `isSameTypedNil` gate can't silently reintroduce
the panic.

**Finding 2 — README `AssertErrorIs`/`AssertNotErrorIs` bullet:
reworded.**
Changed "presence/absence checking with `errors.Is`" to
"presence/absence via `errors.Is`", dropping the dangling gerund and
matching the parallel `errors.As` and `IsErrorFn` bullets. The list
entries are intentionally noun fragments, so I kept the fragment form
rather than expanding one bullet into a full sentence.

For completeness, the earlier DeepSource redundant-type finding on
`errorIs(target error, err error)` is already fixed, folded into
*feat: match errors by their own Is method in IsError*.

`make all test` is green.

@coderabbitai review

**Comment by amery** [comment]
Released in [v0.20.0](https://github.com/darvaza-proxy/core/releases/tag/v0.20.0).

```sh
go get darvaza.org/core@v0.20.0
```

Heads-up for upgraders: the minimum Go version is now 1.25, and
`ErrInvalid` is an alias of `fs.ErrInvalid` — same message, but
`errors.Is` now matches in both directions.
